New Column module for one-dimensional axial force analysis

lfverlaan opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #28053, #28065, #28117, 

#### Brief description of what is fixed or changed
A new class is built for analyzing the influence of axial forces on a one-dimensional element. Functions have been added for applying supports and adding point- and distributed loads. There is also a function for applying telescope hinges, hinges that have a degree of freedom in the axial direction. The load equation is automatically constructed, and it can be solved using a solver function. Lastly, functions have been constructed for defining and plotting the axial force and deflection equations.

#### Other comments
This pull request is built upon pull request #28066. The changes are highlighted in lfverlaan#4.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics
  * Added a new `Column` class for modeling axially loaded structural elements, similar in structure to the `Beam` class.
  * Supports applying axial point loads and fixed horizontal supports using singularity functions.
  * Implemented support for distributed axial loads (with optional `end` location) via singularity function expansion.
  * Added functionality to remove previously applied loads (`remove_load`), updating the internal load equation accordingly.
  * Automatically assembles the total axial load equation symbolically using singularity functions.
  * Introduced a solver method (`solve_for_reaction_loads`) to compute unknown reaction forces by applying equilibrium and boundary conditions.
  * Added support for telescope hinges, allowing displacement in the axial direction.
  * Enabled definition and symbolic solution of axial force and deflection functions.
  * Added plotting capabilities for axial force and deflection diagrams, supporting visualization of internal responses along the column.
<!-- END RELEASE NOTES -->

lfverlaan committed:
New files

Continuum mechanics: New files

Two new files have been added, column.py and a test file for the column script.

lfverlaan committed:
Update .mailmap

lfverlaan committed:
Update test_column.py

Tests have been written for the initialization of the column module, and the formulation of the load equation.

lfverlaan committed:
Update test_column.py

Adjusted the tests. Added a test to show that adding three supports is also compatible. Also added a test to check if reaction forces due to supports are also added to the load equation. Also removed some tests, inputs that are outside of the scope.

lfverlaan committed:
Update column.py

The base of the Column class has been written. A column with a certain length, area and elastic modulus can be initialized. Methods have also been added to apply supports and forces. These methods add the singularity functions to the load equations. The user can get the load equation through the 'load' property.

Tom-van-Woudenberg committed:
Add docs

lfverlaan committed:
Update column.py

A little bit of code cleanliness, removed whitespace

lfverlaan committed:
Merge pull request #1 from Tom-van-Woudenberg/docs

Add docs

lfverlaan committed:
Update column.py

Fixed a mistake where the reaction forces got added to the applied loads list. Now they are simply added to the load equation instead of the apply_load() function being called.

Also added examples in the docstring of the functions.

lfverlaan committed:
Update column.py

Checked the spelling, neatness.

Tom-van-Woudenberg committed:
Fix error in docs

Fixes Title overline too short

lfverlaan committed:
Merge pull request #2 from Tom-van-Woudenberg/docs

Fix error in docs

lfverlaan committed:
Update column.py

removed trailing whitespace

lfverlaan committed:
Update test_column.py

Added tests for the remove load function and the addition of higher order loads.

lfverlaan committed:
Update test_column.py

Adjusted the tests of the test_column to have an input for new variable 'end'. Also fixed some incorrect load equestions

lfverlaan committed:
test_column

Added tests for the solve_for_reaction_loads function. There are tests for statically determnate and indeterminate columns.

lfverlaan committed:
Update column.py

Methods for removing a load and solving the reaction forces have been constructed. Distributed loads have also been integrated, using a helper function called _taylor_helper().

lfverlaan committed:
Update test_column.py

Tests are added for the telescope hinge and equation plotting functions. The telescope hinge function yields incorrect values, and will be updated later.

lfverlaan committed:
Update column.py

A start for the implementation of telescope hinges has been applied. Properties for the reactions loads and hinge deflections have been added, which store the solved values.

Lastly, functions for getting the axial force and deflections expressions have been made. Also functions for plotting these expressions have been set up.

lfverlaan committed:
Adjusted tests for telescope hinges

The last tests for the telescope hinge implementation have been corrected.

lfverlaan committed:
Fixed telescope singularity function

The telescope hinge function now provides the correct singularity function.

The last examples and docstrings have also been added.

lfverlaan commented:
The telescope hinges are now correctly implemented. The first version of the column module is now complete and can be reviewed.

@moorepants @Tom-van-Woudenberg

lfverlaan committed:
Spelling and whitespace

Tom-van-Woudenberg commented:
> The telescope hinges are now correctly implemented. The first version of the column module is now complete and can be reviewed.

@moorepants , could you look at this one first? https://github.com/sympy/sympy/pull/28049. That one has been properly reviewed by me and @saiudayagiri (GSOC).

@saiudayagiri, want to look at this one too?

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_column.py:
```diff
@@ -0,0 +1,475 @@
+    # Test applied loads
+    p = c.applied_loads
+    q = [(10, 0, -1, None), (-10, 10, -1, None)]
+    assert p == q
```
This doesn't test that the load is removed right? Shouldn't you only have the second assert?

Tom-van-Woudenberg replied:
Same for some of the other asserts in this functino

lfverlaan replied:
> This doesn't test that the load is removed right? Shouldn't you only have the second assert?

I guess it is extra, I thought it nicely highlighted the difference after removing the load. Indeed only the remove function should be tested, so the first assert should be left out.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_column.py:
```diff
@@ -0,0 +1,475 @@
+
+test_distributed_loads()
+
+def test_remove_load():
```
Could you test this symbolically too?

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_column.py:
```diff
@@ -0,0 +1,475 @@
+
+test_column()
+
+def test_distributed_loads():
```
Could you test this symbolically too?

Tom-van-Woudenberg reviewed (commented):
I think these release notes should only contain the addition by this PR, not from #28066 

Could you indicate where you deviate from what is implemented in the beam class?

Not for all tests you test them symbolically. @moorepants , is that required?

lfverlaan committed:
Mistake in taylor application

Adjusted a mistake in the taylor helper. The residues of higher order loads weren't being removed properly

Also updated some tests and added more symbolic tests.

lfverlaan committed:
code quality

lfverlaan committed:
code quality

lfverlaan committed:
Code quality

lfverlaan committed:
Cody quality

lfverlaan committed:
Removing whitespaces

lfverlaan committed:
Code quality (got to catch all the whitespace)

lfverlaan committed:
Added blank line at end

lfverlaan committed:
Tests code quality

lfverlaan committed:
Please no more whitespace

lfverlaan committed:
removed unused packages

lfverlaan committed:
Added Tom to mailmap

lfverlaan committed:
Removed excessive returns

lfverlaan committed:
Updated incorrect test

lfverlaan committed:
Removed unneccessary import

lfverlaan committed:
Fix errors in docstrings

saiudayagiri commented on sympy/physics/continuum_mechanics/column.py:
```diff
@@ -0,0 +1,650 @@
+        """
+        This method solves the horizontal reaction loads.
+
+        Returns
```
This method just solves for the reactions right? doesnot return anything

lfverlaan commented on sympy/physics/continuum_mechanics/column.py:
```diff
@@ -0,0 +1,650 @@
+        reaction_solutions = solution[:num_supports]
+        self._reaction_loads = dict(zip(self._applied_supports, reaction_solutions))
+
+        displacement_solutions = solution[num_supports:-1]
```
You are correct, thanks!

lfverlaan committed:
Fixed a few small coding mistakes

saiudayagiri commented:
> > The telescope hinges are now correctly implemented. The first version of the column module is now complete and can be reviewed.
> 
> @moorepants , could you look at this one first? #28049. That one has been properly reviewed by me and @saiudayagiri (GSOC).
> 
> @saiudayagiri, want to look at this one too?



> > The telescope hinges are now correctly implemented. The first version of the column module is now complete and can be reviewed.
> 
> @moorepants , could you look at this one first? #28049. That one has been properly reviewed by me and @saiudayagiri (GSOC).
> 
> @saiudayagiri, want to look at this one too?



> > The telescope hinges are now correctly implemented. The first version of the column module is now complete and can be reviewed.
> 
> @moorepants , could you look at this one first? #28049. That one has been properly reviewed by me and @saiudayagiri (GSOC).
> 
> @saiudayagiri, want to look at this one too?

Reviewed, I added comments and they’ve been resolved—looks good so far! @Tom-van-Woudenberg

lfverlaan committed:
New test

Added test for functionality of having the start be larger than the end of distributed load

lfverlaan committed:
Added remove test

Added remove test for removing a ramp load if start > end

lfverlaan committed:
Updated adding and removing load

Added a case where start > end.

lfverlaan committed:
Fixed test mistake

Tom-van-Woudenberg commented:
> when defining a distributed load from left to right

You mean from right to left?

> Maybe this could be an issue for the beam?

yes, that seems appropriate! I think it's a bug, not an addition.

Tom-van-Woudenberg commented:
@moorepants or @Upabjojr, could you merge this. It's fully reviewed (partly in #28049 and #28066, the failing tests there are fixed in this PR). @saiudayagiri uses this for his GSOC project.

Tom-van-Woudenberg commented:
@lfverlaan , could you update the release notes here with the release notes of https://github.com/sympy/sympy/pull/28049 and https://github.com/sympy/sympy/pull/28066?

lfverlaan commented:
@Tom-van-Woudenberg I changed the title and references. I added all release note points from the previous pull requests. Is this good?

Tom-van-Woudenberg commented:
> @Tom-van-Woudenberg I changed the title and references. I added all release note points from the previous pull requests. Is this good?

Yes, now it's really ready to be merged

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/column.py:
```diff
@@ -0,0 +1,658 @@
+from sympy.series import limit
+from sympy.solvers import linsolve
+
+class Column:
```
It matches the beam class: https://github.com/sympy/sympy/blob/master/sympy/physics/continuum_mechanics/beam.py#L45
